File: Permutations.py
from typing import List

class Solution:

    def permute(self, nums: List[int]) -> List[List[int]]:

            res = []

            def dfs(start, res):
                if start == len(nums):
                    res.append(nums.copy())
                    return
                for i in range(start, len(nums)):
                    nums[start], nums[i] = nums[i], nums[start]
                    dfs(start + 1, res)
                    nums[start], nums[i] = nums[i], nums[start]
            dfs(0, res)
            return res
                
                



    def permute_with_space(self, nums: List[int]) -> List[List[int]]:
        
        map = {}
        for i in nums:
            map[i] = 0
    
        print(nums)

        def dfs(nums, map, lst, res):
            
            if len(lst) == len(nums):
                res.append(lst.copy())

            for i in range(len(nums)):
                if map[nums[i]] != 1:
                    lst.append(nums[i])
                    map[nums[i]] = 1
                    dfs(nums, map, lst, res)
                    map[nums[i]] = 0
                    lst.pop()
                    

        res = []
        dfs(nums, map, [], res)
        return res

nums = [1,2,3]
print(Solution().permute(nums))

File: Permutations.java
import java.util.*;

class Permutations {

    public static void main(String[] args) {
        int[] nums = {1,2,3};
        Permutations obj = new Permutations();
        
        System.out.println(obj.permute(nums));
    }

    public List<List<Integer>> permute(int[] nums) {
        
        class DFS {
            public static void dfs(int start, int[] nums, List<List<Integer>> res) {
                if (start == nums.length) {
                    List<Integer> list = new ArrayList<>();
                    for (int num : nums) {
                        list.add(num);
                    }
                    res.add(new ArrayList<>(list));
                    return;                }

                for (int i = start; i < nums.length; i++) {
                    int temp = nums[i];
                    nums[i] = nums[start];
                    nums[start] = temp;
                    dfs(start + 1, nums, res);
                    temp = nums[i];
                    nums[i] = nums[start];
                    nums[start] = temp;
                }
            }
        }

        List<List<Integer>> res = new ArrayList<>();
        DFS.dfs(0, nums, res);


        return res;
    }

    public List<List<Integer>> permute1(int[] nums) {
        HashMap<Integer, Integer> map = new HashMap<>();

        class DFS {
            public void dfs(int[] nums, HashMap<Integer, Integer> map, List<Integer> lst, List<List<Integer>> res) {
                if (lst.size() == nums.length) {
                    res.add(new ArrayList<>(lst));
                    return;
                }

                for (int i = 0; i < nums.length; i++) {
                    if (!map.containsKey(nums[i])) {
                        map.put(nums[i], 1);
                        lst.add(nums[i]);
                        dfs(nums, map, lst, res);
                        map.remove(nums[i]);
                        lst.remove(lst.size() - 1);
                    }
                }
            }
        }

        

        List<List<Integer>> res = new ArrayList<>();
        DFS obj = new DFS();
        obj.dfs(nums, map, new ArrayList<>(), res);
        return res;
        
    }

}
